// ng-openapi-gen/lib/gen-utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.syncDirs = exports.deleteDirRecursive = exports.resolveRef = exports.tsType = exports.escapeId = exports.serviceClass = exports.modelClass = exports.tsComments = exports.toBasicChars = exports.fileName = exports.methodName = exports.enumName = exports.typeName = exports.ensureNotReserved = exports.namespace = exports.modelFile = exports.qualifiedName = exports.unqualifiedName = exports.simpleName = exports.HTTP_METHODS = void 0;
const fs_extra_1 = __importDefault(require("fs-extra"));
const jsesc_1 = __importDefault(require("jsesc"));
const lodash_1 = require("lodash");
const path_1 = __importDefault(require("path"));
exports.HTTP_METHODS = ['get', 'put', 'post', 'delete', 'options', 'head', 'patch', 'trace'];
/**
 * Returns the simple name, that is, the last part after '/'
 */
function simpleName(name) {
    const pos = name.lastIndexOf('/');
    return name.substring(pos + 1);
}
exports.simpleName = simpleName;
/**
 * Returns the unqualified model class name, that is, the last part after '.'
 */
function unqualifiedName(name, options) {
    const pos = name.lastIndexOf('.');
    return modelClass(name.substring(pos + 1), options);
}
exports.unqualifiedName = unqualifiedName;
/**
 * Returns the qualified model class name, that is, the camelized namespace (if any) plus the unqualified name
 */
function qualifiedName(name, options) {
    const ns = namespace(name);
    const unq = unqualifiedName(name, options);
    return ns ? typeName(ns) + unq : unq;
}
exports.qualifiedName = qualifiedName;
/**
 * Returns the filename where to write a model with the given name
 */
function modelFile(name, options) {
    let result = '';
    const ns = namespace(name);
    if (ns) {
        result += `/${ns}`;
    }
    const file = unqualifiedName(name, options);
    return result += '/' + fileName(file);
}
exports.modelFile = modelFile;
/**
 * Returns the namespace path, that is, the part before the last '.' split by '/' instead of '.'.
 * If there's no namespace, returns undefined.
 */
function namespace(name) {
    name = name.replace(/^\.+/g, '');
    name = name.replace(/\.+$/g, '');
    const pos = name.lastIndexOf('.');
    return pos < 0 ? undefined : name.substring(0, pos).replace(/\./g, '/');
}
exports.namespace = namespace;
const RESERVED_KEYWORDS = ['abstract', 'arguments', 'await', 'boolean', 'break', 'byte', 'case', 'catch', 'char', 'class', 'const', 'continue', 'debugger', 'default', 'delete', 'do', 'double', 'else', 'enum', 'eval', 'export', 'extends', 'false', 'final', 'finally', 'float', 'for', 'function', 'goto', 'if', 'implements', 'import', 'in', 'instanceof', 'int', 'interface', 'let', 'long', 'native', 'new', 'null', 'package', 'private', 'protected', 'public', 'return', 'short', 'static', 'super', 'switch', 'synchronized', 'this', 'throw', 'throws', 'transient', 'true', 'try', 'typeof', 'var', 'void', 'volatile', 'while', 'with', 'yield'];
/**
 * If the given name is a JS reserved keyword, suffix it with a `$` character
 */
function ensureNotReserved(name) {
    return RESERVED_KEYWORDS.includes(name) ? name + '$' : name;
}
exports.ensureNotReserved = ensureNotReserved;
/**
 * Returns the type (class) name for a given regular name
 */
function typeName(name) {
    return (0, lodash_1.upperFirst)(methodName(name));
}
exports.typeName = typeName;
/**
 * Returns the name of the enum constant for a given value
 */
function enumName(value, options) {
    let name = toBasicChars(value, true);
    if (options.enumStyle === 'ignorecase') {
        return name;
    }
    else if (options.enumStyle === 'upper') {
        name = (0, lodash_1.upperCase)(name).replace(/\s+/g, '_');
    }
    else {
        name = (0, lodash_1.upperFirst)((0, lodash_1.camelCase)(name));
    }
    if (/^\d/.test(name)) {
        name = '$' + name;
    }
    return name;
}
exports.enumName = enumName;
/**
 * Returns a suitable method name for the given name
 * @param name The raw name
 */
function methodName(name) {
    return (0, lodash_1.camelCase)(toBasicChars(name, true));
}
exports.methodName = methodName;
/**
 * Returns the file name for a given type name
 */
function fileName(text) {
    return (0, lodash_1.kebabCase)(toBasicChars(text));
}
exports.fileName = fileName;
/**
 * Converts a text to a basic, letters / numbers / underscore representation.
 * When firstNonDigit is true, prepends the result with an uderscore if the first char is a digit.
 */
function toBasicChars(text, firstNonDigit = false) {
    text = (0, lodash_1.deburr)((text || '').trim());
    text = text.replace(/[^\w$]+/g, '_');
    if (firstNonDigit && /[0-9]/.test(text.charAt(0))) {
        text = '_' + text;
    }
    return text;
}
exports.toBasicChars = toBasicChars;
/**
 * Returns the TypeScript comments for the given schema description, in a given indentation level
 */
function tsComments(description, level, deprecated) {
    const indent = '  '.repeat(level);
    if (description === undefined || description.length === 0) {
        return indent + (deprecated ? '/** @deprecated */' : '');
    }
    const lines = description.trim().split('\n');
    let result = '\n' + indent + '/**\n';
    lines.forEach(line => {
        result += indent + ' *' + (line === '' ? '' : ' ' + line.replace(/\*\//g, '* / ')) + '\n';
    });
    if (deprecated) {
        result += indent + ' *\n' + indent + ' * @deprecated\n';
    }
    result += indent + ' */\n' + indent;
    return result;
}
exports.tsComments = tsComments;
/**
 * Applies the prefix and suffix to a model class name
 */
function modelClass(baseName, options) {
    return `${options.modelPrefix || ''}${typeName(baseName)}${options.modelSuffix || ''}`;
}
exports.modelClass = modelClass;
/**
 * Applies the prefix and suffix to a service class name
 */
function serviceClass(baseName, options) {
    return `${options.servicePrefix || ''}${typeName(baseName)}${options.serviceSuffix || 'Service'}`;
}
exports.serviceClass = serviceClass;
/**
 * Escapes the name of a property / parameter if not valid JS identifier
 */
function escapeId(name) {
    if (/^[a-zA-Z]\w*$/.test(name)) {
        return name;
    }
    else {
        return `'${name.replace(/\'/g, '\\\'')}'`;
    }
}
exports.escapeId = escapeId;
/**
 * Appends | null to the given type
 */
function maybeAppendNull(type, nullable) {
    if (` ${type} `.includes('null') || !nullable) {
        // The type itself already includes null
        return type;
    }
    return (type.includes(' ') ? `(${type})` : type) + (nullable ? ' | null' : '');
}
function rawTsType(schema, options, openApi, container) {
    // An union of types
    const union = schema.oneOf || schema.anyOf || [];
    if (union.length > 0) {
        if (union.length > 1) {
            return `(${union.map(u => tsType(u, options, openApi, container)).join(' | ')})`;
        }
        else {
            return union.map(u => tsType(u, options, openApi, container)).join(' | ');
        }
    }
    const type = schema.type || 'any';
    // An array
    if (type === 'array' || schema.items) {
        const items = schema.items || {};
        const itemsType = tsType(items, options, openApi, container);
        return `Array<${itemsType}>`;
    }
    // All the types
    const allOf = schema.allOf || [];
    let intersectionType = [];
    if (allOf.length > 0) {
        intersectionType = allOf.map(u => tsType(u, options, openApi, container));
    }
    // An object
    if (type === 'object' || schema.properties) {
        let result = '{\n';
        const properties = schema.properties || {};
        const required = schema.required;
        for (const baseSchema of allOf) {
            const discriminator = tryGetDiscriminator(baseSchema, schema, openApi);
            if (discriminator) {
                result += `'${discriminator.propName}': '${discriminator.value}';\n`;
            }
        }
        for (const propName of Object.keys(properties)) {
            const property = properties[propName];
            if (!property) {
                continue;
            }
            if (property.description) {
                result += tsComments(property.description, 0, property.deprecated);
            }
            result += `'${propName}'`;
            const propRequired = required && required.includes(propName);
            if (!propRequired) {
                result += '?';
            }
            const propertyType = tsType(property, options, openApi, container);
            result += `: ${propertyType};\n`;
        }
        if (schema.additionalProperties) {
            const additionalProperties = schema.additionalProperties === true ? {} : schema.additionalProperties;
            result += `[key: string]: ${tsType(additionalProperties, options, openApi, container)};\n`;
        }
        result += '}';
        intersectionType.push(result);
    }
    if (intersectionType.length > 0) {
        return intersectionType.join(' & ');
    }
    // Inline enum
    const enumValues = schema.enum || [];
    if (enumValues.length > 0) {
        if (type === 'number' || type === 'integer' || type === 'boolean') {
            return enumValues.join(' | ');
        }
        else {
            return enumValues.map(v => `'${(0, jsesc_1.default)(v)}'`).join(' | ');
        }
    }
    // A Blob
    if (type === 'string' && schema.format === 'binary') {
        return 'Blob';
    }
    // A simple type (integer doesn't exist as type in JS, use number instead)
    return type === 'integer' ? 'number' : type;
}
/**
 * Returns the TypeScript type for the given type and options
 */
function tsType(schemaOrRef, options, openApi, container) {
    if (!schemaOrRef) {
        // No schema
        return 'any';
    }
    if (schemaOrRef.$ref) {
        // A reference
        const resolved = resolveRef(openApi, schemaOrRef.$ref);
        const name = simpleName(schemaOrRef.$ref);
        // When referencing the same container, use its type name
        return maybeAppendNull((container && container.name === name) ? container.typeName : qualifiedName(name, options), !!resolved.nullable);
    }
    // Resolve the actual type (maybe nullable)
    const schema = schemaOrRef;
    const type = rawTsType(schema, options, openApi, container);
    return maybeAppendNull(type, !!schema.nullable);
}
exports.tsType = tsType;
/**
 * Resolves a reference
 * @param ref The reference name, such as #/components/schemas/Name, or just Name
 */
function resolveRef(openApi, ref) {
    if (!ref.includes('/')) {
        ref = `#/components/schemas/${ref}`;
    }
    let current = null;
    for (let part of ref.split('/')) {
        part = part.trim();
        if (part === '#' || part === '') {
            current = openApi;
        }
        else if (current == null) {
            break;
        }
        else {
            current = current[part];
        }
    }
    if (current == null || typeof current !== 'object') {
        throw new Error(`Couldn't resolve reference ${ref}`);
    }
    return current;
}
exports.resolveRef = resolveRef;
/**
 * Recursively deletes a directory
 */
function deleteDirRecursive(dir) {
    if (fs_extra_1.default.existsSync(dir)) {
        fs_extra_1.default.readdirSync(dir).forEach((file) => {
            const curPath = path_1.default.join(dir, file);
            if (fs_extra_1.default.lstatSync(curPath).isDirectory()) { // recurse
                deleteDirRecursive(curPath);
            }
            else { // delete file
                fs_extra_1.default.unlinkSync(curPath);
            }
        });
        fs_extra_1.default.rmdirSync(dir);
    }
}
exports.deleteDirRecursive = deleteDirRecursive;
/**
 * Synchronizes the files from the source to the target directory. Optionally remove stale files.
 */
function syncDirs(srcDir, destDir, removeStale, logger) {
    fs_extra_1.default.ensureDirSync(destDir);
    const srcFiles = fs_extra_1.default.readdirSync(srcDir);
    const destFiles = fs_extra_1.default.readdirSync(destDir);
    for (const file of srcFiles) {
        const srcFile = path_1.default.join(srcDir, file);
        const destFile = path_1.default.join(destDir, file);
        if (fs_extra_1.default.lstatSync(srcFile).isDirectory()) {
            // A directory: recursively sync
            syncDirs(srcFile, destFile, removeStale, logger);
        }
        else {
            // Read the content of both files and update if they differ
            const srcContent = fs_extra_1.default.readFileSync(srcFile, { encoding: 'utf-8' });
            const destContent = fs_extra_1.default.existsSync(destFile) ? fs_extra_1.default.readFileSync(destFile, { encoding: 'utf-8' }) : null;
            if (srcContent !== destContent) {
                fs_extra_1.default.writeFileSync(destFile, srcContent, { encoding: 'utf-8' });
                logger.debug('Wrote ' + destFile);
            }
        }
    }
    if (removeStale) {
        for (const file of destFiles) {
            const srcFile = path_1.default.join(srcDir, file);
            const destFile = path_1.default.join(destDir, file);
            if (!fs_extra_1.default.existsSync(srcFile) && fs_extra_1.default.lstatSync(destFile).isFile()) {
                fs_extra_1.default.unlinkSync(destFile);
                logger.debug('Removed stale file ' + destFile);
            }
        }
    }
}
exports.syncDirs = syncDirs;
/**
 * Tries to get a discriminator info from a base schema and for a derived one.
 */
function tryGetDiscriminator(baseSchemaOrRef, derivedSchema, openApi) {
    var _a;
    const baseSchema = (baseSchemaOrRef.$ref ? resolveRef(openApi, baseSchemaOrRef.$ref) : baseSchemaOrRef);
    const discriminatorProp = (_a = baseSchema.discriminator) === null || _a === void 0 ? void 0 : _a.propertyName;
    if (discriminatorProp) {
        const discriminatorValue = tryGetDiscriminatorValue(baseSchema, derivedSchema, openApi);
        if (discriminatorValue) {
            return {
                propName: discriminatorProp,
                value: discriminatorValue
            };
        }
    }
    return undefined;
}
/**
 * Tries to get a discriminator value from a base schema and for a derived one.
 */
function tryGetDiscriminatorValue(baseSchema, derivedSchema, openApi) {
    var _a, _b;
    const mapping = (_a = baseSchema.discriminator) === null || _a === void 0 ? void 0 : _a.mapping;
    if (mapping) {
        const mappingIndex = Object.values(mapping).findIndex((ref) => resolveRef(openApi, ref) === derivedSchema);
        return (_b = Object.keys(mapping)[mappingIndex]) !== null && _b !== void 0 ? _b : null;
    }
    return null;
}
